File: src/app/core/models/service-request.ts
export type RequestUrgency = 'NORMAL' | 'URGENT';
export type RequestStatus =
  | 'CREATED'
  | 'MATCHING'
  | 'ASSIGNED'
  | 'IN_PROGRESS'
  | 'DONE'
  | 'CANCELED';

export type ContactPref = 'WHATSAPP' | 'CALL' | 'CHAT';

export interface RequestUserLite {
  id: number;
  first_name: string;
  last_name: string;
  email: string;
  role: string;
  is_active: boolean;
}

export interface ServiceRequestCreate {
  category: string;
  title: string;
  description: string;

  urgency?: RequestUrgency;

  city?: string | null;
  neighborhood?: string | null;
  address?: string | null;
  address_ref?: string | null;

  lat?: number | null;
  lng?: number | null;
  accuracy_m?: number | null;

  schedule_date?: string | null;
  time_window?: string | null;

  budget_min?: number | null;
  budget_max?: number | null;

  contact_name?: string | null;
  contact_phone?: string | null;
  contact_pref?: ContactPref | null;
}

export interface ServiceRequest {
  id: number;
  user_id: number;
  assigned_worker_id?: number | null;

  category: string;
  title: string;
  description: string;

  urgency: RequestUrgency;

  city?: string | null;
  neighborhood?: string | null;
  address?: string | null;
  address_ref?: string | null;

  lat?: number | null;
  lng?: number | null;
  accuracy_m?: number | null;

  schedule_date?: string | null;
  time_window?: string | null;

  budget_min?: number | null;
  budget_max?: number | null;

  contact_name?: string | null;
  contact_phone?: string | null;
  contact_pref?: ContactPref | null;

  status: RequestStatus;

  assigned_at?: string | null;
  accepted_at?: string | null;
  started_at?: string | null;
  completed_at?: string | null;

  created_at?: string | null;
  updated_at?: string | null;

  customer?: RequestUserLite | null;
  assigned_worker?: RequestUserLite | null;

  client_name?: string | null;
  client_email?: string | null;
  assigned_worker_name?: string | null;
  assigned_worker_email?: string | null;
}

export type ServiceRequestListItem = ServiceRequest;

export interface RequestMessageCreate {
  body: string;
}

export interface RequestMessage {
  id: number;
  request_id: number;
  sender_user_id: number;
  body: string;
  created_at: string;
  read_at?: string | null;

  sender?: RequestUserLite | null;

  sender_name?: string | null;
  sender_role?: string | null;
  is_mine?: boolean;
}

export interface RequestEvent {
  id: number;
  request_id: number;
  actor_user_id?: number | null;
  event_type: string;
  title: string;
  message?: string | null;
  status_from?: string | null;
  status_to?: string | null;
  meta?: Record<string, any>;
  created_at: string;
  actor?: RequestUserLite | null;
}

File: backend/app/models/service_request.py
from __future__ import annotations

from datetime import datetime
from enum import Enum as PyEnum

from sqlalchemy import (
    Column,
    Integer,
    String,
    Text,
    Date,
    DateTime,
    Enum,
    ForeignKey,
    Float,
    BigInteger,
)
from sqlalchemy.orm import relationship

from ..core.database import Base


class RequestStatus(str, PyEnum):
    CREATED = "CREATED"
    MATCHING = "MATCHING"
    ASSIGNED = "ASSIGNED"
    IN_PROGRESS = "IN_PROGRESS"
    DONE = "DONE"
    CANCELED = "CANCELED"


class RequestUrgency(str, PyEnum):
    NORMAL = "NORMAL"
    URGENT = "URGENT"


class ContactPref(str, PyEnum):
    WHATSAPP = "WHATSAPP"
    CALL = "CALL"
    CHAT = "CHAT"


class ServiceRequest(Base):
    __tablename__ = "service_requests"

    id = Column(Integer, primary_key=True, index=True)

    user_id = Column(
        Integer,
        ForeignKey("users.id", ondelete="CASCADE"),
        nullable=False,
        index=True,
    )

    assigned_worker_id = Column(
        Integer,
        ForeignKey("users.id", ondelete="SET NULL"),
        nullable=True,
        index=True,
    )

    category = Column(String(60), nullable=False, default="GENERAL")
    title = Column(String(120), nullable=False)
    description = Column(Text, nullable=False)

    urgency = Column(Enum(RequestUrgency), nullable=False, default=RequestUrgency.NORMAL)

    city = Column(String(80), nullable=True)
    neighborhood = Column(String(120), nullable=True)
    address = Column(String(200), nullable=True)
    address_ref = Column(String(200), nullable=True)

    lat = Column(Float, nullable=True)
    lng = Column(Float, nullable=True)
    accuracy_m = Column(Integer, nullable=True)

    schedule_date = Column(Date, nullable=True)
    time_window = Column(String(40), nullable=True)

    budget_min = Column(BigInteger, nullable=True)
    budget_max = Column(BigInteger, nullable=True)

    contact_name = Column(String(120), nullable=True)
    contact_phone = Column(String(40), nullable=True)
    contact_pref = Column(Enum(ContactPref), nullable=True)

    status = Column(Enum(RequestStatus), nullable=False, default=RequestStatus.CREATED)

    assigned_at = Column(DateTime, nullable=True)
    accepted_at = Column(DateTime, nullable=True)
    started_at = Column(DateTime, nullable=True)
    completed_at = Column(DateTime, nullable=True)

    created_at = Column(DateTime, nullable=False, default=datetime.utcnow)
    updated_at = Column(
        DateTime,
        nullable=False,
        default=datetime.utcnow,
        onupdate=datetime.utcnow,
    )

    user = relationship(
        "User",
        foreign_keys=[user_id],
        back_populates="requests",
    )

    assigned_worker = relationship(
        "User",
        foreign_keys=[assigned_worker_id],
        back_populates="assigned_requests",
    )

    messages = relationship(
        "RequestMessage",
        back_populates="request",
        cascade="all, delete-orphan",
        passive_deletes=True,
        order_by="RequestMessage.created_at.asc()",
    )

    events = relationship(
        "RequestEvent",
        back_populates="request",
        cascade="all, delete-orphan",
        passive_deletes=True,
        order_by="RequestEvent.created_at.desc()",
    )

    reviews = relationship(
        "RequestReview",
        back_populates="request",
        cascade="all, delete-orphan",
        passive_deletes=True,
        order_by="RequestReview.created_at.desc()",
    )
